PioLSP

anurag3301 opened the pull request:
The `PioLSP` command replaces `Piodb` command to initialise lsp config for chosen lsp server

anurag3301 committed:
feat (PioLSP command to populate lsp config files for a project)

dhupee commented on lua/platformio/piolsp.lua:
no check whether you were in project or not? if not that init command will litter the poor cwd

also in piodebug you have checks if pio is installed or not, not apply that here also?

anurag3301 committed:
show message to restart lsp after PioLSP is ran

anurag3301 commented:
Yes, but this branch will be in progress. Are you working on translating the bash script which you mentioned in #38, if not let me know ill work on it in this branch itself and merge everything later.

anurag3301 commented:
In my test this script has worked nice, it just translate the `.ccls` file to `compile_command.json`. Can you test this script in your workflow. I'll just reimplement it in lua.
```py
import json
from pathlib import Path
import subprocess

# Constants
PROJECT_ROOT = Path.cwd()
SRC_FILE = PROJECT_ROOT / "src" / "main.cpp"
CCLS_FILE = PROJECT_ROOT / ".ccls"
CCDB_NEW = PROJECT_ROOT / "compile_commands.json"

KEEP_FLAGS = ("-I", "-D", "-std", "-Wall")

def parse_ccls():
    common_flags = []
    cpp_flags = []
    compiler = ""
    with open(CCLS_FILE) as f:
        compiler = f.readline().strip()
        for line in f:
            line = line.strip()
            if not line or line.startswith("#"):
                continue
            if line.startswith("%cpp"):
                cpp_flags.extend([flag for flag in line.split()[1:] if flag == "-Wall" or flag.startswith(KEEP_FLAGS)])
            elif line.startswith("%c"):
                pass  # skip C files for now
            elif not line.startswith("%"):
                common_flags.extend([flag for flag in line.split() if flag == "-Wall" or flag.startswith(KEEP_FLAGS)])
    return [compiler] + common_flags + cpp_flags

def generate_single_entry_compile_command(flags):
    entry = {
        "directory": str(PROJECT_ROOT),
        "file": str(SRC_FILE),
        "command": f"{' '.join(flags)}"
    }
    CCDB_NEW.write_text(json.dumps([entry], indent=2))
    return 1

filtered_flags = parse_ccls()
entry_count = generate_single_entry_compile_command(filtered_flags)

entry_count
```

anurag3301 committed:
feat: Generate compile_command.json file from existing .ccls file

anurag3301 committed:
refactored code to use standard global variables

anurag3301 committed:
feat: Adding compile_commands.json to gitignore if its not there

anurag3301 committed:
refactor: project wide stylua execution

anurag3301 committed:
restart LSP after config generation

anurag3301 committed:
ignoring .ccls along with compile_commands.json

anurag3301 commented:
Try with following
```
[env:esp32dev]
platform = espressif32
board = esp32dev
framework = arduino
```

and then do `pio run -t compiledb`, Now if you open a cpp file you see following. 

<img width="1316" height="297" alt="image" src="https://github.com/user-attachments/assets/36cb63d9-e6ff-404c-a53c-56dc7822013c" />


The problem persists with the script given by you. Check if you are able to reproduce this.

anurag3301 committed:
removing unnecessary lsp config files

dhupee commented:
yeah, do as you please, again its your repo

anurag3301 committed:
Enable pioLSP project wide

Now pioinit and piolib triggers piolsp to take care of lsp generation

anurag3301 committed:
fix: enabled appropriate escape characters for build flag

dhupee commented:
what happened?

anurag3301 commented:
Just being little busy, I was testing it and also making a new intro video. The current one is pretty old. Then ill merge it

anurag3301 committed:
Fixes #36: Added user config debug to enable or disable debug output

anurag3301 committed:
Merged main

anurag3301 committed:
Fix: Let user pick source betweek .ccls or compiledb for compile_commands.json when clangd chose as lsp

anurag3301 committed:
Doc: Complete explation for clangd_source

anurag3301 committed:
Fix: recreating on_close callback for each ToggleTerm invok to respect resetLSP

anurag3301 committed:
Update demo video in README.md
